### ISSUE TITLE
[BUG] Commit status check associated with the wrong commit in CircleCI + GitHub combination

### ISSUE THREAD

**Issue by valscion** [danger/danger-js#1423]
**Describe the bug**

Danger associates the commit status check with the latest commit in an open GitHub pull request.

This is problematic when the pull request has more Danger runs and the later Danger run completes faster than the original Danger run. This makes the entire pull request status checks red if Danger has failures.

The commit where Danger did complain on purpose had these environment variables:

```
Using build environment variables:
  BASH_ENV=/tmp/.bash_env-65bb930c74d4b3238d055ed2-0-build
  CI=true
  CIRCLECI=true
  CIRCLE_BRANCH=depfu/update/rubocop-1.60.2
  CIRCLE_BUILD_NUM=467700
  CIRCLE_BUILD_URL=https://circleci.com/gh/venuu/venuu/467700
  CIRCLE_JOB=checks
  CIRCLE_NODE_INDEX=0
  CIRCLE_NODE_TOTAL=1
  CIRCLE_PROJECT_REPONAME=venuu
  CIRCLE_PROJECT_USERNAME=venuu
  CIRCLE_PULL_REQUEST=https://github.com/venuu/venuu/pull/11423
  CIRCLE_PULL_REQUESTS=https://github.com/venuu/venuu/pull/11423
  CIRCLE_REPOSITORY_URL=git@github.com:venuu/venuu.git
  CIRCLE_SHA1=41ee21fb710533f380322143e0b83eace85215ac
  CIRCLE_SHELL_ENV=/tmp/.bash_env-65bb930c74d4b3238d055ed2-0-build
  CIRCLE_USERNAME=Venuusaur
  CIRCLE_WORKFLOW_ID=ee20da3d-45cb-43da-b145-bcbb9b3627b0
  CIRCLE_WORKFLOW_JOB_ID=d7c3eecb-a9ea-4467-a9f4-d4193408f7c9
  CIRCLE_WORKFLOW_WORKSPACE_ID=ee20da3d-45cb-43da-b145-bcbb9b3627b0
  CIRCLE_WORKING_DIRECTORY=~/venuu
  CI_PULL_REQUEST=https://github.com/venuu/venuu/pull/11423
```

The commit where Danger didn't complain but which completed sooner than the one that did complain:

```
Using build environment variables:
  BASH_ENV=/tmp/.bash_env-65bb9358f228fb22b6c5e750-0-build
  CI=true
  CIRCLECI=true
  CIRCLE_BRANCH=depfu/update/rubocop-1.60.2
  CIRCLE_BUILD_NUM=467715
  CIRCLE_BUILD_URL=https://circleci.com/gh/venuu/venuu/467715
  CIRCLE_JOB=checks
  CIRCLE_NODE_INDEX=0
  CIRCLE_NODE_TOTAL=1
  CIRCLE_PROJECT_REPONAME=venuu
  CIRCLE_PROJECT_USERNAME=venuu
  CIRCLE_PULL_REQUEST=https://github.com/venuu/venuu/pull/11423
  CIRCLE_PULL_REQUESTS=https://github.com/venuu/venuu/pull/11423
  CIRCLE_REPOSITORY_URL=git@github.com:venuu/venuu.git
  CIRCLE_SHA1=c995a322ff042af4574820728638a5450eed8cbf
  CIRCLE_SHELL_ENV=/tmp/.bash_env-65bb9358f228fb22b6c5e750-0-build
  CIRCLE_USERNAME=valscion
  CIRCLE_WORKFLOW_ID=9c2dd38a-d5d7-4e91-bf20-be114c9134aa
  CIRCLE_WORKFLOW_JOB_ID=6de8b27a-33a4-45a8-8d5e-25e6541e6a2e
  CIRCLE_WORKFLOW_WORKSPACE_ID=9c2dd38a-d5d7-4e91-bf20-be114c9134aa
  CIRCLE_WORKING_DIRECTORY=~/venuu
  CI_PULL_REQUEST=https://github.com/venuu/venuu/pull/11423
```

**To Reproduce**

1. Setup CircleCI to run Danger
2. Create two commits: One that has a Danger failure but is slow to run and one that doesn't fail and is quick to run
3. Create a GitHub PR with only the slow and failing commit pushed first
4. Push the second commit immediately once the first commit starts to run in CI
5. See that the second commit gets the failing status check even though the first commit was failing

**Expected behavior**

Danger should associate the GitHub commit status with the commit that was failing and not with the latest commit in the PR.

If possible, Danger should use the `CIRCLE_SHA1` environment variable to associate the status check with the correct commit.

**Screenshots**

![image](https://github.com/venuu/venuu/assets/482561/61625d7b-3be1-4fc9-90ce-b8bf5c2baacf)

![image](https://github.com/venuu/venuu/assets/482561/a3bb21a6-7592-4c57-aad5-5576716f1302)

**Your Environment**

 | software         | version
| ---------------- | -------
| danger.js          | 11.3.1
| node             |
| npm              |
| Operating System |

**Additional context**

N/A

**Comment by valscion** [comment]
I wonder if the call in here is missing the last optional argument depicting what the commit hash is?

https://github.com/danger/danger-js/blob/5edb114d07d57292c707bba1dd72b129edd2f44f/source/runner/Executor.ts#L363

Looking at what this code could use:

https://github.com/danger/danger-js/blob/5edb114d07d57292c707bba1dd72b129edd2f44f/source/platforms/github/GitHubAPI.ts#L384-L390

**Comment by fbartho** [comment]
@valscion We’re always happy to review PRs fixing issues. This is a collaborative open-source project!

**Comment by daichunghy** [comment]
Your reproduction matches a concern I am testing in PatchGate: a status written against the latest PR SHA can survive a newer push and then be mistaken for evidence about the current tip. I model `testedSha` explicitly and reject evidence from the wrong revision. Did Danger/CircleCI eventually resolve this by passing the CI SHA into the status API, or is there still a limitation here? I am mainly trying to keep the remediation precise.

**Comment by valscion** [comment]
Yeah I don't know anymore — and as we no longer use CircleCI, this issue is not a priority anymore on me to attempt a fix for. So if someone else were to want to take a chance at this, by all means do.

**Comment by daichunghy** [comment]
Got it. I'll treat the wrong-SHA status as a still-open class of bug, not a CircleCI-only one.

The useful part is that the status API can take the SHA and people still forget to pass it.
